### ISSUE TITLE
drbdsetup failure causes replace-disks to run forever

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#870]
Originally reported of Google Code with ID 814.

```
What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".
# gnt-cluster --version
gnt-cluster (ganeti v2.10.1) 2.10.1
# gnt-cluster version
Software version: 2.10.1
Internode protocol: 2100000
Configuration format: 2100000
OS api version: 20
Export interface: 0
VCS version: (ganeti) version v2.10.1
# hspace --version
hspace (ganeti) version v2.10.1
compiled with ghc 6.12
running on linux x86_64

<b>What distribution are you using?</b>
Debian squeeze

<b>What steps will reproduce the problem?</b>
1. Run 'gnt-instance replace-disks -I hail' on an instance
2. A failure causes the 'drbdsetup syncer' command on the primary to run for a very long time (bug in DRBD)

<b>What is the expected output? What do you see instead?</b>
Expected: The "Changing drbd configuration" step of the replace-disks will time out waiting for drbdsetup, at which point the job will fail. The "Sync devices" step will not happen.
Instead: There is a timeout waiting for drbdsetup, but then the job goes into the "Sync devices" stage. Since the syncer is not set up properly, we see endless output of 0.00% done (like in https://code.google.com/p/ganeti/issues/detail?id=792). Other jobs that attempt to query this instance are blocked by this one.

<b>Please provide any additional information below.</b>

Job log:
# jw 122644
Output from job 122644 follows
------------------------------
Thu Apr 24 17:03:04 2014 Replacing disk(s) 0 for instance '<instance>'
Thu Apr 24 17:03:04 2014 Current primary node: <primary>
Thu Apr 24 17:03:04 2014 Current seconary node: <oldsecondary>
Thu Apr 24 17:03:05 2014 STEP 1/6 Check device existence
Thu Apr 24 17:03:05 2014  - INFO: Checking disk/0 on <primary>
Thu Apr 24 17:03:05 2014  - INFO: Checking volume groups
Thu Apr 24 17:03:05 2014 STEP 2/6 Check peer consistency
Thu Apr 24 17:03:05 2014  - INFO: Checking disk/0 consistency on node <primary>
Thu Apr 24 17:03:05 2014 STEP 3/6 Allocate new storage
Thu Apr 24 17:03:05 2014  - INFO: Adding new local storage on <newsecondary> for disk/0
Thu Apr 24 17:03:08 2014 STEP 4/6 Changing drbd configuration
Thu Apr 24 17:03:08 2014  - INFO: activating a new drbd on <newsecondary> for disk/0
Thu Apr 24 17:03:11 2014  - INFO: Shutting down drbd for disk/0 on old node
Thu Apr 24 17:03:12 2014  - INFO: Detaching primary drbds from the network (=> standalone)
Thu Apr 24 17:03:12 2014  - INFO: Updating instance configuration
Thu Apr 24 17:03:13 2014  - INFO: Attaching primary drbds to new secondary (standalone => connected)
Thu Apr 24 17:18:13 2014  - WARNING: Can't attach drbd disks on node <primary>: Error 28: Operation timed out after 900064 milliseconds with 0 bytes received
Thu Apr 24 17:18:13 2014       Hint: please do a gnt-instance info to see the status of disks
Thu Apr 24 17:18:13 2014 STEP 5/6 Sync devices
Thu Apr 24 17:18:13 2014  - INFO: Waiting for instance <instance> to sync disks
Thu Apr 24 17:18:14 2014  - INFO: - device disk/0:  0.00% done, no time estimate
Thu Apr 24 17:18:14 2014  - INFO: - device disk/0:  0.00% done, no time estimate
... repeats many more times

/var/log/ganeti/master-daemon.log:
2014-04-24 17:18:13,753: ganeti-masterd pid=2617/Jq21/Job122644/I_REPLACE_DISKS ERROR RPC error in drbd_attach_net on node 7558ff92-219a-480b-887a-59d7d6c1b4c3: Error 28: Operation timed out after 900064 milliseconds with 0 bytes received
2014-04-24 17:18:13,754: ganeti-masterd pid=2617/Jq21/Job122644/I_REPLACE_DISKS WARNING Can't attach drbd disks on node <primary>: Error 28: Operation timed out after 900064 milliseconds with 0 bytes received
2014-04-24 17:18:13,934: ganeti-masterd pid=2617/Jq21/Job122644/I_REPLACE_DISKS INFO Waiting for instance <instance> to sync disks
2014-04-24 17:18:14,260: ganeti-masterd pid=2617/Jq21/Job122644/I_REPLACE_DISKS INFO - device disk/0:  0.00% done, no time estimate
2014-04-24 17:18:14,459: ganeti-masterd pid=2617/Jq21/Job122644/I_REPLACE_DISKS INFO - device disk/0:  0.00% done, no time estimate
... repeats many more times
```
Originally added on 2014-04-24 18:17:30 +0000 UTC.

Mentioned Issue with GoogleCode ID 792 has been migrated to GitHub with ID #848.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-04-24 18:28:51 +0000 UTC.

Added Labels: Type-Defect Priority-High
Changed State: Accepted
Added to Milestone: Release2.10

**Comment by Ganeti-Issues-Migrator** [comment]
```
I've got one question: Was the `drbdsetup syncer` command actually killed on the new secondary? The logs you posted only state an RPC timeout, but it would be interesting to see what the noded logs on the new secondary contain (and if `drbdsetup syncer` still ran on the secondary). 

We could (rather easily) just error out of the LU if attaching the new secondary fails. This would let you in the following situation:
 * The instance has no redundancy any more
 * The old secondary node still contains the old secondary disks
 * The new secondary node already contains the new secondary disks, but with no data
This situation would have to be cleaned up manually then (`gnt-instance activate-disks` and removing the stale LVMs, I guess). We have to make sure this doesn't break instances with multiple disks though. Would that be OK?

I see two alternatives:
 * Try to clean up the situation in Ganeti, by either:
  - imposing a timeout on `drbdsetup syncer` and retry the call if it fails (it should be quick anyway). But if there is something wrong with the DRBD module, this wouldn't change a lot...
  - roll back to the old secondary node which still holds the old secondary disk. This sounds rather tricky and risky to get right, especially in the case of multiple disks and a good reason to move the secondary away from the old node.
 * Don't do anything and rely on admins to figure out what went wrong and make them fix the `drbdsetup syncer` problem on the fly

Both options don't sound too compelling to me. So I'd go with the approach above, so just erroring out instead of issuing a warning. What do you think?
```
Originally added on 2014-04-25 11:46:26 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Both the primary and new secondary ran 'drbdsetup syncer' in the pattern of commands shown in [1]. On the new secondary the RPC completed (I see 'POST /drbd_attach_net HTTP/1.1 200' after the commands in the log) but on the primary the drbdsetup command hung forever, leading to the RPC timeout.

For this particular DRBD issue, retrying the drbdsetup syncer call will not help and fixing the drbdsetup syncer problem on the fly isn't really feasible - there's another drbdsetup command running and holding the lock on that device, and as far as we know, the node needs to be rebooted in order to clear the deadlock. Rolling back to the old secondary would certainly be nice, but it does sound complex, and may not even be possible due to the deadlock on the primary. What I'd expect to happen is that the job would observe that the disks are not correctly set up and actively syncing (whether due to the RPC timeout or by gathering some information after that point I'm not sure), and treat that as a failure of the ganeti job, erroring out and releasing locks so that an admin can run other jobs to clean up the cluster. If I understand your comment correctly, that's the solution you're leaning towards as well.

[1]
2014-04-24 17:03:13,960: ganeti-noded pid=1661 INFO RunCmd drbdsetup /dev/drbd0 syncer <params>
2014-04-24 17:03:13,968: ganeti-noded pid=1661 INFO RunCmd drbdsetup /dev/drbd0 net <params>
2014-04-24 17:03:13,977: ganeti-noded pid=1661 INFO RunCmd drbdsetup /dev/drbd0 show
2014-04-24 17:03:14,030: ganeti-noded pid=1661 INFO RunCmd drbdsetup /dev/drbd0 syncer <params>
```
Originally added on 2014-04-25 16:48:22 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fixed with:

commit d33d00ba505c3fdbe7004461ad98a5f06d29c1a8
Author: Thomas Thrainer <thomasth@google.com>
Date:   Mon May 5 09:40:52 2014 +0200

    Fail in replace-disks if attaching disks fails
    
    Previously, if attaching the new secondary during a replace-disks
    operations failed, only a warning was emitted. The subsequent sync-disks
    operation cannot finish in such a case, however.
    
    Therefore, this patch changes the warning into an error. This way it's
    possible for administrators to fix the problem and the cluster does not
    remain in a blocked state.
    
    This fixes issue 814.
    
    Signed-off-by: Thomas Thrainer <thomasth@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>
```
Originally added on 2014-05-09 06:28:27 +0000 UTC.

Mentioned Issue with GoogleCode ID 814 has been migrated to GitHub with ID #870.
Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released in 2.10.4.
```
Originally added on 2014-05-15 10:36:57 +0000 UTC.

Changed State: Released
